### ISSUE TITLE
Github Actions deployment with Google Artifact Registry

### ISSUE THREAD

**Issue by elsmorian** [GoogleCloudPlatform/gcr-cleaner#112]
### TL;DR

Trying to clean a Google Artifact Registry repo from a Github Actions run seems to fail with permissions errors. I have triple check the service account used has both `roles/storage.objectViewer` and `roles/artifactregistry.repoAdmin` however the following are still seen.

My repo is called "images". I added in an extra Github Actions step to try and pull images from the repo, and that succeeds within the run, so the service account has access to pull images etc.

### Expected behavior

GAR repo to be cleaned

### Observed behavior

Stated error

### Debug log output

```text
From the GHActions run:

europe-docker.pkg.dev/gcr-cleaner/gcr-cleaner/gcr-cleaner-cli -recursive -dry-run -keep=10 -repo=europe-west2-docker.pkg.dev/my-gcp-project/images -grace=48h
***"message":"gathering child repositories recursively","severity":"DEBUG","time":"2022-10-23T21:53:46Z"***
***"message":"finding all child repositories","roots":["europe-west2-docker.pkg.dev/my-gcp-project/images"],"severity":"DEBUG","time":"2022-10-23T21:53:46Z"***
***"message":"listing child repositories for registry","registry":"europe-west2-docker.pkg.dev","severity":"DEBUG","time":"2022-10-23T21:53:46Z"***
failed to fetch catalog for registry "europe-west2-docker.pkg.dev": GET https://europe-west2-docker.pkg.dev/v2/_catalog?n=1000: DENIED: The caller does not have permission.
```

### Additional information

Using the following GH Actions workflow:

```yaml
name: 'gcr-cleaner'

on:
  schedule:
    - cron: '0 0 */1 * *' # runs daily
  workflow_dispatch: # allows for manual invocation

jobs:
  gcr-cleaner:
    runs-on: 'ubuntu-latest'
    steps:
      # configure based on your registry
      - name: Login to Artifact Registry
        uses: 'docker/login-action@v2'
        with:
          registry: europe-west2-docker.pkg.dev/my-gcp-project/images
          username: _json_key
          password: ${{ secrets.DOCKER_REGISTRY_PASSWORD }}
      - name: Test step
        shell: bash
        run: 'docker pull europe-west2-docker.pkg.dev/my-gcp-project/images/my_app:current-release'
      # customize based on the gcr-cleaner flags
      - uses: 'docker://europe-docker.pkg.dev/gcr-cleaner/gcr-cleaner/gcr-cleaner-cli'
        env:
          GCRCLEANER_LOG: debug
        with:
          args: >-
            -recursive
            -dry-run
            -keep=10
            -repo=europe-west2-docker.pkg.dev/my-gcp-project/images
            -grace=48h
```

**Comment by elsmorian** [comment]
`GET https://europe-west2-docker.pkg.dev/v2/_catalog?n=1000` looks a bit odd to me, shouldn't the URL be something more like `https://europe-west2-docker.pkg.dev/my-gcp-project/images/v2/_catalog?n=1000`? No idea, just a thought.

**Comment by sethvargo** [comment]
Hi @elsmorian 

Where is `roles/storage.objectViewer` defined? Does this reproduce if you use the service account key json defined in `DOCKER_REGISTRY_PASSWORD` locally with gcr-cleaner-cli?

This is described under the `recursive` section, but that URL is correct: 

>  If the authenticated principal has access to many Container Registry or Artifact Registry repos, this will be very slow! This is because the Docker v2 API does not support server-side filtering, meaning GCR Cleaner must download a manifest of all repositories to which you have access and then do client-side filtering. The most granular filter is at the host layer, meaning GCR Cleaner will perform a list operation on gcr.io (for Container Registry) or us-docker.pkg.dev (for Artifact Registry), parse the response and do client-side filtering to match against the provided patterns, then start deleting. To re-iterate, this operation is not segmented by project - if the authenticated principal has access to 10,000 repos, the client will need to filter through 10,000 repos. The easiest way to mitigate this is to practice the Principle of Least Privilege and create a dedicated service account that has granular permissions on a subset of repositories.

**Comment by elsmorian** [comment]
Hi @sethvargo thanks for the reply!

`roles/storage.objectViewer` is defined on the service account in the `my-gcp-project` project

Unfortunately I can't get it to run locally either, I've tried using my own authed gcloud which is the owner of the project and has full permissions (I can run things like `gcloud artifacts docker tags list` no problem) but have the same error.

```
docker run -e GCRCLEANER_LOG=debug -v "${HOME}/.config/gcloud:/.config/gcloud" -it us-docker.pkg.dev/gcr-cleaner/gcr-cleaner/gcr-cleaner-cli -dry-run -recursive -repo=europe-west2-docker.pkg.dev/my-gcp-project/images
{"message":"gathering child repositories recursively","severity":"DEBUG","time":"2022-10-24T21:02:47Z"}
{"message":"finding all child repositories","roots":["europe-west2-docker.pkg.dev/my-gcp-project/images"],"severity":"DEBUG","time":"2022-10-24T21:02:47Z"}
{"message":"listing child repositories for registry","registry":"europe-west2-docker.pkg.dev","severity":"DEBUG","time":"2022-10-24T21:02:47Z"}
failed to fetch catalog for registry "europe-west2-docker.pkg.dev": GET https://europe-west2-docker.pkg.dev/v2/_catalog?n=1000: DENIED: The caller does not have permission.
```

Understood about the URL, that makes more sense now, thanks!

**Comment by sethvargo** [comment]
Hi @elsmorian 

To run locally, try:

```text
GOOGLE_APPLICATION_CREDENTIALS=/path/to/credentials.json GCR_CLEANER_LOG=debug gcr-cleaner-cli -dry-run -recursive -repo=europe-west2-docker.pkg.dev/my-gcp-project/images
```

There are pre-compiled binaries on [GitHub Releases](https://github.com/GoogleCloudPlatform/gcr-cleaner/releases/tag/v0.9.0-beta1).

**Comment by elsmorian** [comment]
@sethvargo Ahh, hadn't realised you had the binaries in the releases- thanks for the pointer.

Success! That works- when pointing to the service accounts keyfile locally with the binary, It happily found all the repos and iterated through them all.

So maybe something wrong with the docker login action step?

**Comment by sethvargo** [comment]
I think you're hitting https://github.com/GoogleCloudPlatform/gcr-cleaner/issues/101, but it's been fixed since August...

**Comment by elsmorian** [comment]
Ah! I did try pinning the image to `gcr-cleaner-cli:0.9.0-beta1'` just incase it wasn't pulling the latest, but that hasn't helped

**Comment by sethvargo** [comment]
Could you try using `google-github-actions/auth` instead of `docker-login` to see if that works?

```yaml
steps:
- uses: 'actions/checkout@v3'

- id: 'auth'
  uses: 'google-github-actions/auth@v0'
  with:
    credentials_json: '${{ secrets.GOOGLE_CREDENTIALS }}'

- uses: 'docker://europe-docker.pkg.dev/gcr-cleaner/gcr-cleaner/gcr-cleaner-cli'
  env:
    GCRCLEANER_LOG: debug
  with:
    args: >-
      -recursive
      -dry-run
      -keep=10
      -repo=europe-west2-docker.pkg.dev/my-gcp-project/images
      -grace=48h
```

**Comment by elsmorian** [comment]
@sethvargo That worked! Using `google-github-actions/auth@v0` worked staright away, no problems.
